File: packages/api/src/endpoints/epoch.ts
import type { EpochDetailParamResponse, EpochDetailStatsResponse, EpochListResponse } from "@/types/epochTypes";

import { handleFetch } from "@/lib/handleFetch";

/**
 * Fetches the list of epochs from the server.
 *
 * This endpoint returns a list of available epochs in the blockchain,
 * typically including epoch numbers and their metadata.
 *
 * @returns {Promise<EpochListResponse>} A promise resolving to the epoch list data.
 *
 * @throws Will throw an error if the request fails or times out (default timeout: 60 seconds).
 */
export const getEpochList = async () => {
  const url = "/epoch/list";

  const options = {
    timeout: 60000,
  };

  return handleFetch<EpochListResponse>(url, undefined, options);
};

/**
 * Fetches detailed protocol parameters for a specific epoch.
 *
 * @param {number} no - The epoch number to fetch parameters for.
 * @returns {Promise<EpochDetailParamResponse>} A promise resolving to the epoch parameter details.
 *
 * @throws Will throw an error if the request fails or the response is invalid.
 */
export const getEpochDetailParam = async (no: number) => {
  const url = "/epoch/param";
  const options = {
    params: { no },
  };

  return handleFetch<EpochDetailParamResponse>(url, undefined, options);
};

/**
 * Fetches statistical data for a specific epoch.
 *
 * @param {number} no - The epoch number to retrieve statistics for.
 * @returns {Promise<EpochDetailStatsResponse>} A promise resolving to the epoch statistics response.
 *
 * @throws Will throw an error if the fetch request fails or the response is invalid.
 */
export const getEpochDetailStats = async (no: number) => {
  const url = "/epoch/stats";
  const options = {
    params: { no },
  };

  return handleFetch<EpochDetailStatsResponse>(url, undefined, options);
};

File: packages/cexplorer-api-rs/src/endpoints/epoch.rs
use crate::client::{fetch, fetch_with_params};
use crate::error::CexplorerError;
use crate::types::epoch_types::{
    EpochListResponse, EpochDetailParamResponse, EpochDetailStatsResponse,
};
use serde::Serialize;

/// Get list of all epochs
pub async fn get_epoch_list() -> Result<EpochListResponse, CexplorerError> {
    let endpoint = "/epoch/list";
    fetch::<EpochListResponse>(endpoint).await
}

#[derive(Debug, Serialize)]
struct EpochNoParams {
    no: u64,
}

/// Get protocol parameters for a specific epoch
pub async fn get_epoch_detail_param(no: u64) -> Result<EpochDetailParamResponse, CexplorerError> {
    let endpoint = "/epoch/param";
    let params = EpochNoParams { no };
    fetch_with_params::<EpochDetailParamResponse, EpochNoParams>(endpoint, Some(&params)).await
}

/// Get statistics for a specific epoch
pub async fn get_epoch_detail_stats(no: u64) -> Result<EpochDetailStatsResponse, CexplorerError> {
    let endpoint = "/epoch/stats";
    let params = EpochNoParams { no };
    fetch_with_params::<EpochDetailStatsResponse, EpochNoParams>(endpoint, Some(&params)).await
}